src/main/java/org/broadinstitute/hellbender/tools/htsgetreader/HtsgetResponse.java

```diff
@@ -26,17 +46,78 @@
         @JsonProperty("class")
         private HtsgetClass dataClass;
 
+        @JsonIgnore
+        private static CloseableHttpClient client;
+
         public URI getUri() {
             return this.uri;
         }
 
         public Map<String, String> getHeaders() {
-            return this.headers;
+            return Collections.unmodifiableMap(this.headers);
         }
 
         public HtsgetClass getDataClass() {
             return this.dataClass;
         }
+
+        private static synchronized CloseableHttpClient getHttpClient() {
+            if (Block.client == null) {
+                Block.client = HttpClientBuilder.create()
+                    .setConnectionManager(new PoolingHttpClientConnectionManager())
+                    .setServiceUnavailableRetryStrategy(new ServiceUnavailableRetryStrategy() {
+                        @Override
+                        // retry at most 4 times if a 5xx status code is received, or no status line is present
+                        public boolean retryRequest(final HttpResponse resp, final int executionCount, final HttpContext context) {
+                            if (executionCount > 4) {
+                                return false;
+                            }
+                            if (resp.getStatusLine() == null) {
+                                return true;
+                            }
+                            final int statusCode = resp.getStatusLine().getStatusCode();
+                            return 500 <= statusCode && statusCode < 600;
+                        }
+
+                        @Override
+                        public long getRetryInterval() {
+                            return 0;
+                        }
+                    })
+                    .build();
+            }
+            return Block.client;
+        }
+
+        public InputStream getData() {
```

Comment:
Could you add a comment to this method explaining what happens here?  It's useful to know that it potentially gets saved to a temp file and then you read from the tmp file and that's not obvious unless you read the code.

Reply:
Added.